### PR TITLE
Fix cg of admm: warm start, using inversion for non convolution case

### PR DESCRIPTION
**Pull request by EnLAI111**
For non-convolution case, when we set ```tol=1e-12``` for ```cg```, ADMM can converge. But in this case, it is faster to use the inversion of matrix and get the exact solution, instead of solving an equation.

### CODE REVIEW

File: solvers/ADMM.py [hunk 1]
```diff
@@ -38,18 +39,27 @@ def run(self, callback):
...
         mu = np.zeros(p - 1)
         gamma = self.gamma
+        tol_cg = 1e-12
         Aty = self.A.T @ self.y
-        AtA_gDtD = LinearOperator(shape=(p, p),
-                                  matvec=lambda x: self.A.T @ (self.A @ x) -
-                                  gamma * np.diff(np.diff(x),
-                                                  append=0, prepend=0))
-
+        if type(self.A) == np.ndarray:
```
Comment (tomMoral) [inline solvers/ADMM.py:44]: Better to use `isinstance(A, np.ndarray)` as this would work with subclasses.

File: solvers/ADMM.py [hunk 2]
```diff
@@ -38,18 +39,27 @@ def run(self, callback):
...
+                                                        append=0,
+                                                        prepend=0))
         while callback(u):
             z_old = z
             u_tmp = (Aty + np.diff(mu, append=0, prepend=0)
-                     - gamma * np.diff(z, append=0, prepend=0)
-                     )
-            u, _ = cg(AtA_gDtD, u_tmp)
+                     - gamma * np.diff(z, append=0, prepend=0))
+            if type(self.A) == np.ndarray:
```
Comment (agramfort) [inline solvers/ADMM.py:59]:
```suggestion
            if isinstance(self.A, np.ndarray):
```

File: solvers/ADMM.py [hunk 3]
```diff
@@ -38,18 +39,27 @@ def run(self, callback):
...
+                u = np.ravel(AtA_gDtD_inv @ u_tmp)
+            else:
+                u, _ = cg(AtA_gDtD, u_tmp, x0=u, tol=tol_cg)
```
[the remark this answers is not in the document]
Comment (EnLAI111) [reply solvers/ADMM.py:62]: It's a little faster with warm start but not so obvious.

image

**Review by agramfort — APPROVED** [review]
merge if happy and it's green 🙏

**Pull request merged**